### PR TITLE
[SPARK-17358][SQL] Cached table(parquet/orc) should be shard between beelines

### PR DESCRIPTION
**Pull request by watermen**
## What changes were proposed in this pull request?

Cached table(parquet/orc) couldn't be shard between beelines, because the `sameResult` method used by `CacheManager` always return false(`sparkSession` are different) when compare two `HadoopFsRelation` in different beelines. So we make `sparkSession` a curry parameter.
## How was this patch tested?

Beeline1

```
1: jdbc:hive2://localhost:10000> CACHE TABLE src_pqt;
+---------+--+
| Result  |
+---------+--+
+---------+--+
No rows selected (5.143 seconds)
1: jdbc:hive2://localhost:10000> EXPLAIN SELECT * FROM src_pqt;
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--+
|                                                                                                                                                                                                            plan                                                                                                                                                                                                            |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--+
| == Physical Plan ==
InMemoryTableScan [key#49, value#50]
   +- InMemoryRelation [key#49, value#50], true, 10000, StorageLevel(disk, memory, deserialized, 1 replicas), `src_pqt`
         +- *FileScan parquet default.src_pqt[key#0,value#1] Batched: true, Format: ParquetFormat, InputPaths: hdfs://199.0.0.1:9000/qiyadong/src_pqt, PartitionFilters: [], PushedFilters: [], ReadSchema: struct<key:int,value:string>  |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--+
```

Beeline2

```
0: jdbc:hive2://localhost:10000> EXPLAIN SELECT * FROM src_pqt;
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--+
|                                                                                                                                                                                                            plan                                                                                                                                                                                                            |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--+
| == Physical Plan ==
InMemoryTableScan [key#68, value#69]
   +- InMemoryRelation [key#68, value#69], true, 10000, StorageLevel(disk, memory, deserialized, 1 replicas), `src_pqt`
         +- *FileScan parquet default.src_pqt[key#0,value#1] Batched: true, Format: ParquetFormat, InputPaths: hdfs://199.0.0.1:9000/qiyadong/src_pqt, PartitionFilters: [], PushedFilters: [], ReadSchema: struct<key:int,value:string>  |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--+
```

### CODE REVIEW

**Comment by watermen** [comment]
@yhuai @cloud-fan Can you review the code for me?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala [hunk 1]
```diff
@@ -169,6 +169,13 @@ case class HadoopFsRelation(
     location.allFiles().map(_.getPath.toUri.toString).toArray
 
   override def sizeInBytes: Long = location.allFiles().map(_.getLen).sum
+
+  override def equals(other: Any): Boolean = other match {
+    case r: HadoopFsRelation => location == r.location
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala]: Is it safe to completely ignore the other options? Why not ignore just the session and nothing else?

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala]: +1

**Comment by cloud-fan** [comment]
instead of implementing `equals` and `hashCode`, can we just make `sparkSession` a curry parameter? e.g. `case class HadoopFsRelation(location: FileCatalog, ...)(sparkSession: SparkSession)`. If the amount of changes is acceptable, I think it's a better approach.

**Comment by watermen** [comment]
@cloud-fan, if we just make sparkSession a curry parameter, `sparkSession` can't be accessed. e.g.
```scala
val supportsBatch = relation.fileFormat.supportBatch(
  relation.sparkSession, StructType.fromAttributes(output))
```
Or we need to add a method `_sparkSession` in `HadoopFsRelation` to get the value of `sparkSession`?
```scala
def _sparkSession = sparkSession
```

**Comment by cloud-fan** [comment]
oh sorry, like this `case class HadoopFsRelation(location: FileCatalog, ...)(val sparkSession: SparkSession)`

**Comment by cloud-fan** [comment]
LGTM, can you update the description? thanks!

**Comment by watermen** [comment]
@cloud-fan I did, thanks for your advices.

**Comment by cloud-fan** [comment]
thanks, merging to master and 2.0!

**Pull request closed without merging**